### ISSUE TITLE
after marriage proposal PR, redirect the user to aETH token if the current token is ETH and the name of the action is `depositAAVE`

### ISSUE THREAD

**Issue by James-Sangalli** [AlphaWallet/alpha-wallet-ios#2056]
Observe this file https://github.com/AlphaWallet/TokenScript-Examples/blob/train-leaves/examples/action/ETH/ETH.xml
go to line 71 and see the hardcoded name is `depositeAAVE`

-> only PR to `train-leaves`

**Comment by vladyslav-iosdev** [comment]
@hboon i got one question, i added ` Aave Interest bearing ETH (aETH) token` through add custom token screen. When i tap on token it should be displayd in `TokenInstanceViewController` but for me it shows in `TokenViewController`, seems like i added it in wrong way. right?

**Comment by hboon** [comment]
@vladyslav-iosdev nope, it's defined as an ERC20, a fungible, so it shows `TokenViewController` instead of `TokensCardViewController` (in which if you tap 1 of the rows, opens `TokenInstanceViewController`). `TokensCardViewController` is for non-fungibles.

**Comment by hboon** [comment]
@vladyslav-iosdev look at `TokensViewController.didSelect(token:in:)`

**Comment by vladyslav-iosdev** [comment]
ok, got it so these actions i need to add manually? i thought that it will come with actions like for for non-fungible tokens
<img width="115" alt="Screenshot 2020-07-22 at 09 27 22" src="https://user-images.githubusercontent.com/55975226/88142137-a4301380-cbfd-11ea-824b-e335009c660a.png">

**Comment by hboon** [comment]
@vladyslav-iosdev Oh I see. Yes it will be automatic. I just createdPR #2060 which after it is merged into the `train-leaves` branch should show action buttons like you would expect them to. If you are working on this before that PR is merged, you can  clone the `activity` feature branch and then `git rebase train-leaves` on your feature branch once the PR is merged.

**Comment by vladyslav-iosdev** [comment]
@hboon how can walk through `deposit to aave` flow? i got some eth i tap on `deposit to aave` it opens input screen, i tap confirm, and nothing happens

**Comment by hboon** [comment]
@vladyslav-iosdev could be a problem with the TokenScript file or the app itself. @James-Sangalli ?

But the code you want to add/modify is related to #2030.

**Comment by vladyslav-iosdev** [comment]
> But the code you want to add/modify is related to #2030.

yep, i already added it, and now what to make sure that it works correctly.

**Comment by vladyslav-iosdev** [comment]
> @vladyslav-iosdev could be a problem with the TokenScript file or the app itself. @James-Sangalli ?
> 
> But the code you want to add/modify is related to #2030.

it breaks on this line
`guard count == self.count else { return }` counts don't match

**Comment by vladyslav-iosdev** [comment]
@hboon do we have this screen?

<img width="148" alt="Screenshot 2020-07-22 at 11 13 37" src="https://user-images.githubusercontent.com/55975226/88166172-88d60000-cc1f-11ea-8e6b-383d355d6695.png">

**Comment by vladyslav-iosdev** [comment]
also one question: to make sure that it is `depositeAAVE` i need to look at `action.transactionFunction.functionType.functionName` ?

**Comment by hboon** [comment]
>@hboon do we have this screen? ...

@vladyslav-iosdev not yet. I’m working on it. You can pretend that is TokenViewController.

**Comment by hboon** [comment]
>also one question: to make sure that it is `depositeAAVE` i need to look at `action.transactionFunction.functionType.functionName` ? 

@vladyslav-iosdev no. Maybe ‘action.name’? I hope it’s a property.

**Comment by vladyslav-iosdev** [comment]
eth.xml that comes from activity branch and eth.xml from https://github.com/AlphaWallet/TokenScript-Examples/blob/train-leaves/examples/action/ETH/ETH.xml are different. activity. branch version doesn't have depositeAAVE, and even if add it manually in ETH.xml, it doesn't appear in action.

**Comment by vladyslav-iosdev** [comment]
@hboon sorry deleted  your comment by mistake

**Comment by vladyslav-iosdev** [comment]
@hboon according to logic for this task if token is eth i need to check if token is token.type == .nativeCryptocurrency and name == "ETH" ?

**Comment by hboon** [comment]
I think you can check .contract.sameContract(as: Constants.nativeCrypto) and .server == .mainnet

**Comment by hboon** [comment]
And the action name of course.

**Comment by hboon** [comment]
>eth.xml that comes from activity branch and eth.xml from https://github.com/AlphaWallet/TokenScript-Examples/blob/train-leaves/examples/action/ETH/ETH.xml are different. activity. branch version doesn't have depositeAAVE, and even if add it manually in ETH.xml, it doesn't appear in action.

@vladyslav-iosdev I fhink it’s this: https://github.com/AlphaWallet/TokenScript-Examples/blob/train-leaves/examples/action/ETH/ETH.xml

**Comment by hboon** [comment]
The file is embedded in the app. There’s a eth.tsml.

**Comment by vladyslav-iosdev** [comment]
@hboon are u able to deposit eth? does it work for u?
also, am i understand right after we deposition eth, it will create a new aeth/aave token for us, and we need to open TokenViewController for this token, right? 
how could i determine that this token i need has been appeared in token list?

**Comment by hboon** [comment]
> @hboon are u able to deposit eth? does it work for u?
> also, am i understand right after we deposition eth, it will create a new aeth/aave token for us, and we need to open TokenViewController for this token, right?
> how could i determine that this token i need has been appeared in token list?

@James-Sangalli ^

**Comment by hboon** [comment]
> how could i determine that this token i need has been appeared in token list?

But my suggestion for this is. Since this is a special build of the app, just hardcode the aETH token (and token you need) into the list `Constants.partnerContracts` so the app should automatically add it (be sure to test it does show up).

Then when you want to display an instance of `TokenViewController` for it, use `TokensDataStore.token(forContract:) and if it's nil, programatically add it before trying again.

**Comment by vladyslav-iosdev** [comment]
> My suggestion for this is: Since this is a special build of the app, just hardcode the aETH token (and any token you need) into the list Constants.partnerContracts so the app should automatically add it a while after app starts up (be sure to test it does show up) if there's connectivity.
>Then when you want to display an instance of TokenViewController for it, use `TokensDataStore.token(forContract:) and if it's nil, programatically add it before trying again.

ok, thx.

**Comment by vladyslav-iosdev** [comment]
> Then when you want to display an instance of `TokenViewController` for it, use `TokensDataStore.token(forContract:) and if it's nil, programatically add it before trying again.

by `programatically add it before trying again.` what do you mean? we can add token only with address field filled?

**Comment by hboon** [comment]
Look for how `Constants.partnerContracts` is used to see how a token is programmatically added to the database. Basically do that if the token returned is nil (ie. still not in the database)

**Comment by vladyslav-iosdev** [comment]
@James-Sangalli 
related to issue when i can't perform depositing to aave, first its because in function `createPromiseForSubscribeOnce`
i can't reach `seal.fulfill(Void())`, after i update it with `guard count == self.count - 1 else { return } ` for now to pass through it, it started crash at 
`values.merging(tokenLevelAttributeValues.compactMapValues { $0.value.resolvedValue }) { (_, new) in new }`
i updated it with `compactMapValues` and after that i get error  with message
`insufficient funds for gas * price + value`

also i found that attribute `variableInterestRateBorrow` that comes from `tokenLevelAttributeValues` can't be resolved

**Comment by vladyslav-iosdev** [comment]
@hboon or @James-Sangalli u send me aETH to 0xf3e310f7EC8301bb188DF75A451c9d7B1548AfB3 for testing?

**Comment by hboon** [comment]
>@James-Sangalli  ...

@vladyslav-iosdev are you able to proceed without that? If you can, go ahead first and then create an issue pointing me to a commit or branch so I can reproduce it and fix it?

**Comment by vladyslav-iosdev** [comment]
> > @James-Sangalli  ...
> 
> @vladyslav-iosdev are you able to proceed without that? If you can, go ahead first and then create an issue pointing me to a commit or branch so I can reproduce it and fix it?

ok, i will work on the other parts of this task

**Comment by James-Sangalli** [comment]
> @hboon are u able to deposit eth? does it work for u?
> also, am i understand right after we deposition eth, it will create a new aeth/aave token for us, and we need to open TokenViewController for this token, right?
> how could i determine that this token i need has been appeared in token list?

@vladyslav-iosdev if you deposit ETH, it will mint new aETH tokens for you. So long as the transaction is successful it will occur. A MintOnDeposit event is emitted from this transaction.

**Comment by James-Sangalli** [comment]
> @hboon or @James-Sangalli u send me aETH to 0xf3e310f7EC8301bb188DF75A451c9d7B1548AfB3 for testing?

Sent you some, you can also create more by depositing ETH

**Comment by James-Sangalli** [comment]
> i updated it with `compactMapValues` and after that i get error with message

@vladyslav-iosdev please pull the activity branch as this has been fixed.

**Comment by vladyslav-iosdev** [comment]
> > i updated it with `compactMapValues` and after that i get error with message
> 
> @vladyslav-iosdev please pull the activity branch as this has been fixed.



> > @hboon are u able to deposit eth? does it work for u?
> > also, am i understand right after we deposition eth, it will create a new aeth/aave token for us, and we need to open TokenViewController for this token, right?
> > how could i determine that this token i need has been appeared in token list?
> 
> @vladyslav-iosdev if you deposit ETH, it will mint new aETH tokens for you. So long as the transaction is successful it will occur. A MintOnDeposit event is emitted from this transaction.

ok, got it, thanks. how could i handle this transaction event, what class do i need to check ?

**Comment by vladyslav-iosdev** [comment]
@James-Sangalli for now it seems like `enabledForBorrowing` can't be resolved.

**Comment by vladyslav-iosdev** [comment]
@James-Sangalli can you please check #2056 PR, and let me know if it works, i am not able to test this case.

**Comment by James-Sangalli** [comment]
Dropped
